File: src/design_patterns/structural/decorator/c#/decorator.cs
// Component
abstract class Component
{
    public abstract string Operation();
}

// Concrete Component
class ConcreteComponent : Component
{
    public override string Operation()
    {
        return "ConcreteComponent";
    }
}

// Decorator
abstract class Decorator : Component
{
    protected Component _component;

    public Decorator(Component component)
    {
        _component = component;
    }

    public override string Operation()
    {
        return _component.Operation();
    }
}

// Concrete Decorator A
class ConcreteDecoratorA : Decorator
{
    public ConcreteDecoratorA(Component component) : base(component) { }

    public override string Operation()
    {
        return $"ConcreteDecoratorA({base.Operation()})";
    }
}

// Concrete Decorator B
class ConcreteDecoratorB : Decorator
{
    public ConcreteDecoratorB(Component component) : base(component) { }

    public override string Operation()
    {
        return $"ConcreteDecoratorB({base.Operation()})";
    }
}

/*
// Usage
Component component = new ConcreteComponent();
Decorator decorator1 = new ConcreteDecoratorA(component);
Decorator decorator2 = new ConcreteDecoratorB(decorator1);
Console.WriteLine(decorator2.Operation());  // Output: ConcreteDecoratorB(ConcreteDecoratorA(ConcreteComponent))
*/

File: src/design_patterns/structural/decorator/python/decorator.py
from abc import ABC, abstractmethod

# Component
class Component(ABC):
    """ """
    @abstractmethod
    def operation(self) -> str:
        """ """
        pass

# Concrete Component
class ConcreteComponent(Component):
    """ """
    def operation(self) -> str:
        """ """
        return "ConcreteComponent"

# Decorator
class Decorator(Component):
    """ """
    def __init__(self, component: Component) -> None:
        self._component = component

    @abstractmethod
    def operation(self) -> str:
        """ """
        pass

# Concrete Decorator A
class ConcreteDecoratorA(Decorator):
    """ """
    def operation(self) -> str:
        """ """
        return f"ConcreteDecoratorA({self._component.operation()})"

# Concrete Decorator B
class ConcreteDecoratorB(Decorator):
    """ """
    def operation(self) -> str:
        """ """
        return f"ConcreteDecoratorB({self._component.operation()})"

# Usage
component = ConcreteComponent()
decorator1 = ConcreteDecoratorA(component)
decorator2 = ConcreteDecoratorB(decorator1)
print(decorator2.operation())  # Output: ConcreteDecoratorB(ConcreteDecoratorA(ConcreteComponent))
